File: src/claudectl/operations/context.py
"""Context and environment setup operations for workspaces."""

from __future__ import annotations

import shutil
from pathlib import Path

from claudectl.domain.git import get_repo_root


def copy_claude_context(workspace_path: Path, source_root: Path | None = None) -> list[str]:
    """Copy Claude local settings and CLAUDE.md to a workspace.

    Copies files that aren't tracked by git but are needed for Claude
    to have proper context and permissions in the new workspace.

    Files copied (if they exist in source but not in workspace):
    - .claude/settings.local.json
    - CLAUDE.md

    Args:
        workspace_path: Path to the workspace to copy files to
        source_root: Source repository root. If None, uses current repo.

    Returns:
        List of files that were copied
    """
    if source_root is None:
        source_root = get_repo_root()

    copied: list[str] = []

    # Files to copy (relative to repo root)
    files_to_copy = [
        Path(".claude") / "settings.local.json",
        Path("CLAUDE.md"),
    ]

    for rel_path in files_to_copy:
        source_file = source_root / rel_path
        dest_file = workspace_path / rel_path

        # Only copy if source exists and dest doesn't
        if source_file.exists() and not dest_file.exists():
            # Create parent directory if needed
            dest_file.parent.mkdir(parents=True, exist_ok=True)
            shutil.copy2(source_file, dest_file)
            copied.append(str(rel_path))

    return copied

File: internal/context/context.go
// Package context provides utilities for managing context files.
package context //nolint:revive // Package name matches purpose, conflict with stdlib is acceptable

import (
	"io"
	"os"
	"path/filepath"
)

// CopyClaudeContext copies Claude local settings and CLAUDE.md to a workspace.
// Copies files that aren't tracked by git but are needed for Claude
// to have proper context and permissions in the new workspace.
func CopyClaudeContext(workspacePath, sourceRoot string) ([]string, error) {
	var copied []string

	// Files to copy (relative to repo root)
	filesToCopy := []string{
		".claude/settings.local.json",
		"CLAUDE.md",
		".mcp.json",
	}

	for _, relPath := range filesToCopy {
		sourceFile := filepath.Join(sourceRoot, relPath)
		destFile := filepath.Join(workspacePath, relPath)

		// Check if source exists
		if _, err := os.Stat(sourceFile); os.IsNotExist(err) {
			continue
		}

		// Check if dest already exists
		if _, err := os.Stat(destFile); err == nil {
			continue
		}

		// Create parent directory if needed
		if err := os.MkdirAll(filepath.Dir(destFile), 0755); err != nil { //nolint:gosec // Context directories need to be readable
			return nil, err
		}

		// Copy file
		if err := copyFile(sourceFile, destFile); err != nil {
			return nil, err
		}

		copied = append(copied, relPath)
	}

	return copied, nil
}

func copyFile(src, dst string) error {
	source, err := os.Open(src) //nolint:gosec // Path is controlled, opening context files
	if err != nil {
		return err
	}
	defer func() { _ = source.Close() }()

	dest, err := os.Create(dst) //nolint:gosec // Path is controlled, creating context files
	if err != nil {
		return err
	}
	defer func() { _ = dest.Close() }()

	_, err = io.Copy(dest, source)
	if err != nil {
		return err
	}

	// Copy file mode
	sourceInfo, err := source.Stat()
	if err != nil {
		return err
	}
	return os.Chmod(dst, sourceInfo.Mode())
}
